components/brave_perf_predictor/browser/perf_predictor_tab_helper.cc

```diff
@@ -0,0 +1,126 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_perf_predictor/browser/perf_predictor_tab_helper.h"
+
+#include "brave/components/brave_perf_predictor/browser/third_party_extractor.h"
+#include "brave/components/brave_perf_predictor/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+namespace brave_perf_predictor {
+
+PerfPredictorTabHelper::PerfPredictorTabHelper(
+    content::WebContents* web_contents)
+    : WebContentsObserver(web_contents),
+      bandwidth_predictor_(std::make_unique<BandwidthSavingsPredictor>()) {
+  if (web_contents->GetBrowserContext()->IsOffTheRecord())
+    return;
+
+#if BUILDFLAG(BRAVE_P3A_ENABLED)
+  bandwidth_tracker_ = std::make_unique<P3ABandwidthSavingsTracker>(
+      user_prefs::UserPrefs::Get(web_contents->GetBrowserContext()));
+#endif
+}
+
+PerfPredictorTabHelper::~PerfPredictorTabHelper() = default;
+
+void PerfPredictorTabHelper::DidStartNavigation(
+    content::NavigationHandle* handle) {
+  if (!handle || !handle->IsInMainFrame() || handle->IsDownload())
+    return;
+  // Gather prediction of the _previous_ navigation
+  if (navigation_id_ != 0)
+    RecordSavings();
+}
+
+void PerfPredictorTabHelper::ReadyToCommitNavigation(
```

Comment:
i'm pretty sure we can drop this method and move everything meaningful into `DidFinishNavigation`.
Also we probably want to filter out same document navigations? To keep the predictor state in case we just navigate to a different fragment